JAR reading: more accurate and widespread caching, less read contention

retronym opened the pull request:
Multithreaded users of the compiler can experience lock contention in
classpath reading. This occurs because a) scalac by default reuses a
`ZipFile` instance for a given JAR across concurrent compilers to reduce the
memory footprint, and b) because `j.u.ZipFile` contains does not support
concurrent reads.

Instead of having a 1-1 relationship between `ZipArchive` and `j.u.ZipFile`,
this PR uses an internal pool of `ZipFile` instances, which can be expanded
as needed to service concurrent reads a given JAR file.

While investigating this, I noticed that our caching layer was not
considering the value of the `-release` compiler option as part of the
cache key. This is now fixed. I also included Java 9+ classpath reposotories
(`JrtClassPath` and `CtSymClassPath` in the caching scheme.)

Profiling showed my that the attempt to reduce allocations when converting
the classpath entry names into symbol names was counter productive as it
bypassed `lazy val name` in `ClassFileEntryImpl` which was forced anyway on
other code paths.

dwijnand commented:
```
% scalac t6591_3.scala
% <in process execution of run/t6591_3.scala> > t6591_3-run.log
% diff /home/jenkins/workspace/scala-2.12.x-validate-main/test/files/run/t6591_3.check /home/jenkins/workspace/scala-2.12.x-validate-main/test/files/run/t6591_3-run.log
@@ -1 +1,13 @@
 Block(List(ValDef(Modifiers(), TermName("v"), Select(This(TypeName("A")), TypeName("I")), Apply(Select(New(Select(This(TypeName("A")), TypeName("I"))), termNames.CONSTRUCTOR), List()))), Ident(TermName("v")))
+Uncaught exception on thread Thread[Timer-0,5,main]: java.nio.file.NoSuchFileException: t6502-run.obj/test1.jar
+java.nio.file.NoSuchFileException: t6502-run.obj/test1.jar
+	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
+	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
+	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
+	at sun.nio.fs.UnixPath.toRealPath(UnixPath.java:837)
+	at com.sun.nio.zipfs.ZipFileSystemProvider.removeFileSystem(ZipFileSystemProvider.java:322)
+	at com.sun.nio.zipfs.ZipFileSystem.close(ZipFileSystem.java:313)
+	at scala.tools.nsc.classpath.ZipFsClassPath.close(DirectoryClassPath.scala:220)
+	at scala.tools.nsc.classpath.FileBasedCache$$anon$1$$anon$2.run(ZipAndJarFileLookupFactory.scala:249)
+	at java.util.TimerThread.mainLoop(Timer.java:555)
+	at java.util.TimerThread.run(Timer.java:505)
```

retronym commented:
Nothing major other than the risk of new code. I'd like to battle test it to flush out further bugs and refine the performance. 

I'm not sure if it is shipped with non-OpenJDK based distros, so we might need the old implementation as a fallback.

retronym commented:
A clarification: `java.nio.zipfs.ZipFileSystem` does not actually support concurerent reads of _data_, but it does support concurrent reads of _metadata_.

retronym commented:
My earlier analysis was wrong, I do see contention on reads in this implementation, which is visible in my multi-threaded compilation benchmark. The improvement that I thought I saw might have been due to #9549 instead.

retronym commented:
Update:
  - I've tried to make zipfs lock free on reads by tweaking the way it does File I/O. https://github.com/retronym/jdk/pull/2. I'm floating this idea with the JDK team. https://twitter.com/retronym/status/1381823131875086336
  - I've pushed another commit here that uses a pool of plain-old java.util.ZipFiles to avoid the contention but avoid needing to open a ZipFile instance for each class we read.

retronym committed:
Cache JRT and --release (ct.sym) classpath elements.

Classpath elements based on a) jrt:// file system (representing platform libraries of the
current the Java 9+ instance) and b) ct.sym (the JEP 247 repository of the of previous
JDK versions) are an immutable part of the JDK. The ClassPath entries we create are
safe to share across concurrent or subsequent compilers in the same way we cache
entries for regular JARs.

retronym committed:
Include --release version in cache key for classpath cache

This prevents concurrent compilers with different values for this compiler
option from seeing the incorrect API.

retronym committed:
Avoid contention in classpath access

Classpath caching shares a single instance of ZipArchive across
multiple threads. This can cause read contention as j.u.ZipFile
internally serializes reads.

Instead, maintain a pool of ZipFile instances to avoid sharing
them across threads.

retronym committed:
Adjust conversion of AbstractFile's name to TypeName

The current optimized version tries to avoid temporary strings.
But it doesn't achieve this for classes based by jrt:// (or any
`NioPath`, as the call to `AbstractFile.fileName` internally
constructs a string each time.

This commit uses `.name` (which is a `lazy val`).
